File: frontend/src/api/modules/diagnosis.ts
/**
 * 智能诊断 API
 */
import request from '@/utils/request'
import type { PageParams, PageResponse } from './types'

// ==================== 类型定义 ====================

/** 诊断原因条目 */
export interface DiagnosisCause {
  cause: string
  confidence: number
  suggested_actions: string[]
  rule_code?: string
}

/** 诊断规则 */
export interface DiagnosisRule {
  id: number
  rule_code: string
  name: string
  description: string | null
  category: string
  trigger_condition: Record<string, unknown> | null
  diagnosis_logic: Record<string, unknown> | null
  priority: number
  is_enabled: boolean
  is_system: boolean
  created_at: string
  updated_at: string
}

/** 创建诊断规则参数 */
export interface DiagnosisRuleCreate {
  rule_code: string
  name: string
  description?: string
  category: string
  trigger_condition: Record<string, unknown>
  diagnosis_logic: Record<string, unknown>
  priority?: number
  is_enabled?: boolean
}

/** 更新诊断规则参数 */
export interface DiagnosisRuleUpdate {
  name?: string
  description?: string
  category?: string
  trigger_condition?: Record<string, unknown>
  diagnosis_logic?: Record<string, unknown>
  priority?: number
  is_enabled?: boolean
}

/** 诊断结果 */
export interface DiagnosisResult {
  id: number
  alarm_id: number | null
  alarm_no: string | null
  rule_id: number | null
  rule_code: string | null
  device_type: string | null
  zone: string | null
  causes: DiagnosisCause[] | null
  diagnosis_time_ms: number
  created_at: string
}

/** 诊断分类 */
export interface DiagnosisCategory {
  code: string
  name: string
  count: number
}

// ==================== API 函数 ====================

/**
 * 获取诊断规则列表
 */
export function getDiagnosisRules(params?: PageParams & {
  category?: string
  is_enabled?: boolean
  is_system?: boolean
}): Promise<PageResponse<DiagnosisRule>> {
  return request.get('/v1/diagnosis/rules', { params })
}

/**
 * 获取诊断规则详情
 */
export function getDiagnosisRule(id: number): Promise<DiagnosisRule> {
  return request.get(`/v1/diagnosis/rules/${id}`)
}

/**
 * 创建诊断规则
 */
export function createDiagnosisRule(data: DiagnosisRuleCreate): Promise<DiagnosisRule> {
  return request.post('/v1/diagnosis/rules', data)
}

/**
 * 更新诊断规则
 */
export function updateDiagnosisRule(id: number, data: DiagnosisRuleUpdate): Promise<DiagnosisRule> {
  return request.put(`/v1/diagnosis/rules/${id}`, data)
}

/**
 * 删除诊断规则
 */
export function deleteDiagnosisRule(id: number): Promise<{ message: string }> {
  return request.delete(`/v1/diagnosis/rules/${id}`)
}

/**
 * 启用/禁用诊断规则
 */
export function toggleDiagnosisRule(id: number): Promise<DiagnosisRule> {
  return request.put(`/v1/diagnosis/rules/${id}/toggle`)
}

/**
 * 重载 YAML 诊断规则
 */
export function reloadDiagnosisRules(): Promise<{ message: string; count: number }> {
  return request.get('/v1/diagnosis/rules/reload')
}

/**
 * 获取诊断结果列表
 */
export function getDiagnosisResults(params?: PageParams & {
  device_type?: string
  zone?: string
  start_time?: string
  end_time?: string
}): Promise<PageResponse<DiagnosisResult>> {
  return request.get('/v1/diagnosis/results', { params })
}

/**
 * 获取诊断结果详情
 */
export function getDiagnosisResult(id: number): Promise<DiagnosisResult> {
  return request.get(`/v1/diagnosis/results/${id}`)
}

/**
 * 按告警ID查询诊断结果
 */
export function getDiagnosisByAlarm(alarmId: number): Promise<DiagnosisResult[]> {
  return request.get(`/v1/diagnosis/results/by-alarm/${alarmId}`)
}

/**
 * 手动触发诊断
 */
export function manualDiagnose(alarmId: number): Promise<{ message: string; alarm_id: number }> {
  return request.post(`/v1/diagnosis/analyze/${alarmId}`)
}

/**
 * 获取诊断分类列表
 */
export function getDiagnosisCategories(): Promise<DiagnosisCategory[]> {
  return request.get('/v1/diagnosis/categories')
}

// ==================== 标注管理 (Story 24.8) ====================

/** 诊断标注 */
export interface DiagnosisAnnotation {
  id: number
  session_id: number
  annotator_id: number | null
  annotation: 'accurate' | 'inaccurate' | 'unknown'
  actual_root_cause: string | null
  notes: string | null
  annotated_at: string
  created_at: string
  updated_at: string
}

/** 创建标注参数 */
export interface DiagnosisAnnotationCreate {
  session_id: number
  annotation: 'accurate' | 'inaccurate' | 'unknown'
  actual_root_cause?: string
  notes?: string
}

/** 标注统计 */
export interface DiagnosisAnnotationStats {
  total_annotations: number
  accurate_count: number
  inaccurate_count: number
  unknown_count: number
  accurate_rate: number
  user_stats: Array<{
    user_id: number
    username: string
    annotation_count: number
  }>
  top_annotators: Array<{
    user_id: number
    username: string
    annotation_count: number
  }>
}

/**
 * 创建诊断标注
 */
export function createDiagnosisAnnotation(data: DiagnosisAnnotationCreate): Promise<DiagnosisAnnotation> {
  return request.post('/v1/diagnosis/annotations', data)
}

/**
 * 获取标注列表
 */
export function getDiagnosisAnnotations(params?: PageParams & {
  session_id?: number
  annotator_id?: number
  annotation?: string
}): Promise<PageResponse<DiagnosisAnnotation>> {
  return request.get('/v1/diagnosis/annotations', { params })
}

/**
 * 删除标注
 */
export function deleteDiagnosisAnnotation(id: number): Promise<{ message: string }> {
  return request.delete(`/v1/diagnosis/annotations/${id}`)
}

/**
 * 获取标注统计
 */
export function getDiagnosisAnnotationStats(topN?: number): Promise<DiagnosisAnnotationStats> {
  return request.get('/v1/diagnosis/annotations/stats', {
    params: { top_n: topN || 10 }
  })
}

// ==================== 反事实分析 (Story 26.1) ====================

/** 反事实场景 */
export interface CounterfactualScenario {
  removed_evidence_id: number
  new_root_cause: string | null
  new_confidence: number
  confidence_change: number
  conclusion_changed: boolean
}

/** 反事实分析结果 */
export interface CounterfactualAnalysis {
  id: number
  session_id: number
  original_root_cause: string | null
  original_confidence: number
  top_evidences: Array<{
    node_id: number
    evidence_type: string
    probability: number
    sensor_weight: number
    path_length: number
  }>
  analysis_results: CounterfactualScenario[]
  analysis_time_ms: number
  fault_tree_version: string | null
  config_version: string
  created_at: string
}

/**
 * 获取反事实分析结果
 */
export function getCounterfactualAnalysis(sessionId: number): Promise<CounterfactualAnalysis> {
  return request.get(`/v1/diagnosis/counterfactual/${sessionId}`)
}

/**
 * 获取反事实分析列表
 */
export function getCounterfactualAnalysisList(params?: PageParams & {
  min_confidence?: number
  start_date?: string
  end_date?: string
}): Promise<PageResponse<CounterfactualAnalysis>> {
  return request.get('/v1/diagnosis/counterfactual', { params })
}

// ==================== 误诊反馈报告 (Story 26.2) ====================

/** 系统报告 */
export interface SystemReport {
  id: number
  report_type: string
  report_period: string
  report_version: string
  content: string
  summary: Record<string, unknown> | null
  generated_at: string
  generated_by: string | null
  deleted_at: string | null
  updated_at: string
}

/** 报告摘要 */
export interface ReportSummary {
  total_sessions: number
  annotated_count: number
  annotation_coverage: number
  false_positive_count: number
  false_negative_count: number
  accuracy_rate: number | null
}

/**
 * 获取误诊报告列表
 */
export function getMisdiagnosisReports(params?: PageParams & {
  start_period?: string
  end_period?: string
}): Promise<PageResponse<SystemReport>> {
  return request.get('/v1/diagnosis/reports/misdiagnosis', { params })
}

/**
 * 获取误诊报告详情
 */
export function getMisdiagnosisReport(reportId: number): Promise<SystemReport> {
  return request.get(`/v1/diagnosis/reports/misdiagnosis/${reportId}`)
}

/**
 * 手动生成误诊报告
 */
export function generateMisdiagnosisReport(period: string): Promise<{
  report_id: number
  message: string
}> {
  return request.post('/v1/diagnosis/reports/misdiagnosis/generate', null, {
    params: { period }
  })
}

/**
 * 导出误诊报告为 PDF
 */
export function exportMisdiagnosisReport(reportId: number): Promise<Blob> {
  return request.get(`/v1/diagnosis/reports/misdiagnosis/${reportId}/export`, {
    responseType: 'blob'
  })
}

// ==================== 时间窗口调参 API ====================

/** 时间窗口调整记录 */
export interface TimeWindowAdjustment {
  id: number
  device_type: string
  current_window_minutes: number
  proposed_window_minutes: number
  adjustment_percent: number
  sample_count: number
  p50_duration_seconds: number
  p90_duration_seconds: number
  status: string
  reason?: string
  approved_by?: number
  approved_at?: string
  created_at: string
  updated_at: string
}

/** 时间窗口调整查询参数 */
export interface TimeWindowAdjustmentQuery extends PageParams {
  device_type?: string
  status?: string
}

/** 审批/拒绝参数 */
export interface TimeWindowApprovalRequest {
  reason?: string
}

/**
 * 获取时间窗口调整记录列表
 */
export function getTimeWindowAdjustments(
  params: TimeWindowAdjustmentQuery
): Promise<PageResponse<TimeWindowAdjustment>> {
  return request.get('/v1/diagnosis/time-window-tuning/adjustments', { params })
}

/**
 * 触发时间窗口调参分析
 */
export function triggerTimeWindowAnalysis(): Promise<{
  analyzed_device_types: number
  total_adjustments: number
  pending_approvals: number
}> {
  return request.post('/v1/diagnosis/time-window-tuning/analyze')
}

/**
 * 审批时间窗口调整
 */
export function approveTimeWindowAdjustment(
  adjustmentId: number,
  data: TimeWindowApprovalRequest
): Promise<{
  message: string
  device_type: string
  new_window_minutes: number
}> {
  return request.post(`/v1/diagnosis/time-window-tuning/adjustments/${adjustmentId}/approve`, data)
}

/**
 * 拒绝时间窗口调整
 */
export function rejectTimeWindowAdjustment(
  adjustmentId: number,
  data: TimeWindowApprovalRequest
): Promise<{
  message: string
}> {
  return request.post(`/v1/diagnosis/time-window-tuning/adjustments/${adjustmentId}/reject`, data)
}

File: backend/app/schemas/diagnosis.py
"""
诊断 Schema
Story 9-3: 智能故障诊断
Story 24.6: 诊断会话、审计日志、结果扩展
Story 25.3: UPS电池SOH预测
Story 25.4: N+X冗余拓扑与断路器保护逻辑
"""

from datetime import datetime
from typing import Optional, List, Any

from pydantic import BaseModel, ConfigDict, Field, model_validator, field_validator


# ==================== Cause Item ====================


class DiagnosisCauseItem(BaseModel):
    """诊断原因条目"""

    cause: str
    confidence: int
    suggested_actions: List[str] = []


# ==================== Rule Schemas ====================


class DiagnosisRuleCreate(BaseModel):
    """创建诊断规则"""

    rule_code: str
    name: str
    description: Optional[str] = None
    category: str
    trigger_condition: dict
    diagnosis_logic: dict
    priority: int = 0
    is_enabled: bool = True


class DiagnosisRuleUpdate(BaseModel):
    """更新诊断规则"""

    name: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    trigger_condition: Optional[dict] = None
    diagnosis_logic: Optional[dict] = None
    priority: Optional[int] = None
    is_enabled: Optional[bool] = None


class DiagnosisRuleResponse(BaseModel):
    """诊断规则响应"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    rule_code: str
    name: str
    description: Optional[str] = None
    category: str
    trigger_condition: Optional[dict] = None
    diagnosis_logic: Optional[dict] = None
    priority: int
    is_enabled: bool
    is_system: bool
    created_at: datetime
    updated_at: datetime


# ==================== Result Schemas ====================


class DiagnosisResultResponse(BaseModel):
    """诊断结果响应"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    alarm_id: Optional[int] = None
    alarm_no: Optional[str] = None
    rule_id: Optional[int] = None
    rule_code: Optional[str] = None
    device_type: Optional[str] = None
    zone: Optional[str] = None
    causes: Optional[List[DiagnosisCauseItem]] = None
    diagnosis_time_ms: int = 0
    created_at: datetime

    # Story 24.6 扩展字段
    session_id: Optional[int] = None
    root_cause: Optional[str] = None
    confidence: Optional[float] = None
    reasoning_path: Optional[Any] = None
    evidence: Optional[Any] = None
    fault_tree_version: Optional[str] = None


# ==================== Session Schemas ====================


class DiagnosisSessionResponse(BaseModel):
    """诊断会话响应"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    trigger_alarm_id: Optional[int] = None
    device_id: Optional[int] = None
    engine_level: str
    status: str
    push_status: str
    max_confidence: Optional[float] = None
    start_time: datetime
    end_time: Optional[datetime] = None
    inference_time_ms: int = 0
    created_at: datetime
    result: Optional[DiagnosisResultResponse] = None


class DiagnosisSessionListQuery(BaseModel):
    """诊断会话列表查询参数"""

    page: int = Field(default=1, ge=1, description="页码")
    page_size: int = Field(default=20, ge=1, le=100, description="每页数量")
    device_id: Optional[int] = Field(default=None, description="设备ID")
    engine_level: Optional[str] = Field(default=None, description="推理级别: L1/L2/L3")
    min_confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0, description="最低置信度")
    start_date: Optional[datetime] = Field(default=None, description="开始时间")
    end_date: Optional[datetime] = Field(default=None, description="结束时间")


# ==================== Audit Log Schemas ====================


class DiagnosisAuditLogResponse(BaseModel):
    """诊断审计日志响应"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    session_id: int
    input_data: Optional[Any] = None
    output_data: Optional[Any] = None
    engine_level: str
    inference_time_ms: int = 0
    fault_tree_version: Optional[str] = None
    created_at: datetime


# ==================== Category Schema ====================


class DiagnosisCategoryItem(BaseModel):
    """诊断分类条目"""

    code: str
    name: str
    count: int = 0


# ==================== Annotation Schemas ====================


class DiagnosisAnnotationCreate(BaseModel):
    """创建诊断标注"""

    session_id: int = Field(..., description="诊断会话ID")
    annotation: str = Field(..., description="标注结果: accurate/inaccurate/unknown")
    actual_root_cause: Optional[str] = Field(None, max_length=1000, description="实际根因(标注为inaccurate时必填)")
    notes: Optional[str] = Field(None, max_length=2000, description="备注")


class DiagnosisAnnotationResponse(BaseModel):
    """诊断标注响应"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    session_id: int
    annotator_id: Optional[int] = None
    annotation: str
    actual_root_cause: Optional[str] = None
    notes: Optional[str] = None
    annotated_at: datetime
    created_at: datetime
    updated_at: datetime


class DiagnosisAnnotationListQuery(BaseModel):
    """诊断标注列表查询参数"""

    page: int = Field(default=1, ge=1, description="页码")
    page_size: int = Field(default=20, ge=1, le=100, description="每页数量")
    session_id: Optional[int] = Field(default=None, description="会话ID")
    annotator_id: Optional[int] = Field(default=None, description="标注者ID")
    annotation: Optional[str] = Field(default=None, description="标注结果")


class DiagnosisAnnotationStatsResponse(BaseModel):
    """诊断标注统计响应"""

    total_annotations: int = Field(..., description="总标注数")
    accurate_count: int = Field(..., description="准确标注数")
    inaccurate_count: int = Field(..., description="不准确标注数")
    unknown_count: int = Field(..., description="未知标注数")
    accurate_rate: float = Field(..., description="准确率")
    user_stats: List[dict] = Field(..., description="用户标注统计")
    top_annotators: List[dict] = Field(..., description="Top标注者")


# ==================== Battery SOH Schemas ====================


class BatterySOHRecordCreate(BaseModel):
    """创建电池SOH记录"""

    device_id: int
    soh_percent: float = Field(ge=0, le=100, description="SOH百分比 [0-100]")
    resistance_mohm: Optional[float] = Field(default=None, gt=0, description="当前内阻(毫欧)")
    cycle_count: Optional[int] = Field(default=None, ge=0, description="充放电循环次数")
    weights_version: Optional[str] = None


class BatterySOHRecordResponse(BatterySOHRecordCreate):
    """电池SOH记录响应"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    calculated_at: datetime


class SOHWeightsConfig(BaseModel):
    """SOH权重配置"""

    w_r: float = Field(ge=0, le=1, description="内阻权重")
    w_c: float = Field(ge=0, le=1, description="循环次数权重")
    version: str = Field(default="v1.0", description="配置版本")

    @model_validator(mode='after')
    def validate_weights_sum(self):
        """验证权重之和约为 1.0（允许 ±0.1 误差）"""
        total = self.w_r + self.w_c
        if not (0.9 <= total <= 1.1):
            raise ValueError(f"权重之和应约为 1.0，当前为 {total:.2f}")
        return self


# ==================== Breaker Profile Schemas - Story 25.4 ====================


class BreakerProfileCreate(BaseModel):
    """创建断路器配置"""

    breaker_device_id: int = Field(..., description="断路器设备ID")
    trip_curve_type: str = Field(..., description="脱扣曲线类型: B/C/D")
    rated_current: float = Field(..., gt=0, description="额定电流 A")

    @field_validator('trip_curve_type')
    @classmethod
    def validate_trip_curve_type(cls, v: str) -> str:
        """验证脱扣曲线类型"""
        if v not in ('B', 'C', 'D'):
            raise ValueError("脱扣曲线类型必须是 'B', 'C', 或 'D'")
        return v


class BreakerProfileUpdate(BaseModel):
    """更新断路器配置"""

    trip_curve_type: Optional[str] = Field(None, description="脱扣曲线类型: B/C/D")
    rated_current: Optional[float] = Field(None, gt=0, description="额定电流 A")

    @field_validator('trip_curve_type')
    @classmethod
    def validate_trip_curve_type(cls, v: Optional[str]) -> Optional[str]:
        """验证脱扣曲线类型"""
        if v is not None and v not in ('B', 'C', 'D'):
            raise ValueError("脱扣曲线类型必须是 'B', 'C', 或 'D'")
        return v


class BreakerProfileResponse(BaseModel):
    """断路器配置响应"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    breaker_device_id: int
    trip_curve_type: str
    rated_current: float
    created_at: datetime
    updated_at: datetime
